### ISSUE TITLE
How to understand the "warm up" for NATTEN

### ISSUE THREAD

**Issue by fastvideo-anyscale** [SHI-Labs/NATTEN#192]
Hi authors, thanks a lot for your wonderful project, for the following use scenerio,
```

import torch
import natten
natten.use_fused_na(True)
natten.use_kv_parallelism_in_fused_na(True)
from functools import partial
from natten.functional import na3d
import time
batch_size = 1

nheads = 4

headdim = 128

device = 'cuda'
dtype = torch.bfloat16

q, k, v = [torch.randn(batch_size, 32, 48, 80, nheads, headdim, device=device, dtype=dtype,
                    requires_grad=True) for _ in range(3)]

natten_attn = partial(na3d, kernel_size=(17,33,33))
for i in range(50):
    torch.cuda.synchronize()
    start = time.time()
    out = na3d(q, k, v, kernel_size=3)
    end = time.time()
    print(f"forward: {end - start:3f}")

    loss = out.sum()

    torch.cuda.synchronize()
    start = time.time()
    loss.backward()
    torch.cuda.synchronize()
    end = time.time()
    print(f"backward: {end - start:3f}")
```
The time for forward and backward is
```
forward: 0.028309
backward: 0.096288
```
but converge to
```
forward: 0.000199
backward: 0.067154
```
in a super fast way.
Do you have any suggestions to improve the performance further and how to understand the "warm up" for NATTEN, thanks a lot in advance.

**Comment by alihassanijr** [comment]
Thank you for your interest.

The behavior of the script above, as far as I can tell, is not something specific to NATTEN, but rather something that I'd expect from any other operator.

Generally speaking, performance comes with variation, and that's why performance measurements are usually done over a number of iterations. Warmup iterations are usually separated out and not measured by the way.

However, one thing jumps out in the script above, which is that you're calling NATTEN ops on the same tensors over and over again. This can produce numbers that a re a bit misleading, because depending on large your tensors are and the size of your GPU's L2 cache, at some point those tensors may become L2-resident, and you observe a speedup because most of the DRAM accesses are L2 cache hits, therefore you're dealing with a much higher memory bandwidth.

I see you're using a kernel size 3, which is probably almost always memory-bandwidth-bound on modern hardware, so that's probably why there's that speedup. But I wouldn't count on the same speedup to reoccur exactly when you're training an entire model with this layer.

I hope this clarifies things for you, but if not, please let me know.

**Comment by fastvideo-anyscale** [comment]
Hi, Author, thanks a lot for you reply. I have read the code in a more detailed way and your explanation about cache warm up and it helps a lot.
